Add web frontend solidjs

t-animal opened the pull request.

yogan committed:
Add SolidJS project (based on Vite TS template)

yogan committed:
Configure CORS in BE to allow local connections

yogan committed:
Add simple socket.io connection to backend

yogan committed:
Render received messages

t-animal committed:
Use coffee cup instead of solid logo

t-animal committed:
Implement protocol for receiving messages

t-animal committed:
Implement UI and protocol for sending

mxwnk commented on frontend-web/src/protocol.ts:
```diff
@@ -0,0 +1,57 @@
+  name: string;
+}
+
+export class Protocol {
```
Do we want to keep this in a root folder? We could introduce an api folder or something like that

yogan replied:
I can hear @t-animal screaming "told you so!" in the distance.  
It was me who (for now) voted against that with the argument to not over-organize as long as we only have like three real source files.

Are you fine with us keeping everything in `src/` for now, or does it bother you a lot?

mxwnk replied:
Totally fine for me 😊
I already approved this PR, so I don't see a problem, why we couldn't merge it. My comments are only some questions, suggestions and ideas.

mxwnk commented on frontend-web/src/protocol.ts:
```diff
@@ -0,0 +1,57 @@
+        this.socket
+          .messages()
+          .map(this.parseRawMessage)
+          .filter((elem) => elem !== null) as CallMessage[]
```
This cast feels a little bit dirty, why do we need this?

yogan replied:
TypeScript is  not clever enough to narrow down the type from `(CallMessage | null)[]`.

The potential `null`s are coming from `parseRawMessage` and indicate malformed messages.  
Filtering them out is fine, and after the `filter` call, we can be 100% sure to have `CallMessage[]`.

Still, it's a bid sad to have the `as` there.

yogan replied:
Found a better way using a [type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates).

See 381e12c

t-animal replied:
`return 'InvalidMessage';`

Not sure what to think of this. On the one hand I like that this self-describes so nicely. On the other hand, a string could easily be mistaken for a valid value whereas null outright screams "I'm a invalid message"

yogan replied:
We agreed to keep it as it is.

mxwnk commented on backend/src/main.ts:
```diff
@@ -5,7 +5,12 @@ import { createServer } from "http";
-const io = new Server(httpServer);
+const io = new Server(httpServer, {
+    cors: {
+        origin: "http://localhost:3000",
```
Just an idea: but we could switch to https for local dev as well. I was using https://www.npmjs.com/package/vite-plugin-mkcert lately to prevent some dev problems

yogan replied:
Nice idea, I've created #12 so that we don't forget about it.

yogan committed:
Filter out invalid messages /w type predicate

We can get rid of the ugly `as` cast by using a type predicate for the
function passed to `filter`.

See:
https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates
